packages/opentelemetry-scope-async-hooks/src/AsyncHooksScopeManager.ts

```diff
@@ -0,0 +1,158 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ScopeManager } from '@opentelemetry/scope-base';
+import * as asyncHooks from 'async_hooks';
+import { EventEmitter } from 'events';
+import * as shimmer from 'shimmer';
+
+const EVENT_EMITTER_METHODS: Array<keyof EventEmitter> = [
+  'addListener',
+  'on',
+  'once',
+  'prependListener',
+  'prependOnceListener',
+];
+
+export class AsyncHooksScopeManager implements ScopeManager {
+  private asyncHook: asyncHooks.AsyncHook;
+  private scopes: { [uid: number]: unknown } = {};
+  private current: unknown = null;
+
+  constructor() {
+    this.asyncHook = asyncHooks.createHook({
+      init: this.init.bind(this),
+      before: this.before.bind(this),
+      destroy: this.destroy.bind(this),
+    });
+  }
+
+  active(): unknown {
+    return this.current;
+  }
+
+  with<T extends (...args: unknown[]) => ReturnType<T>>(
+    scope: unknown,
+    fn: T
+  ): ReturnType<T> {
+    const oldCurrent = this.current;
+    this.current = scope;
+    const res = fn();
+    this.current = oldCurrent;
+    return res;
+  }
+
+  bind<T>(target: T, scope?: unknown): T {
+    // if no specific scope to propagate is give, we use the current one
+    if (!scope) {
+      scope = this.current;
+    }
+    if (target instanceof EventEmitter) {
+      return this.bindEventEmitter(target, scope);
+    } else if (typeof target === 'function') {
+      return this.bindFunction(target, scope);
+    }
+    return target;
+  }
+
+  enable(): this {
+    this.asyncHook.enable();
+    return this;
+  }
+
+  disable(): this {
+    this.asyncHook.disable();
+    this.scopes = {};
+    this.current = null;
+    return this;
+  }
+
+  private bindFunction<T extends Function>(target: T, scope?: unknown): T {
+    const manager = this;
+    const boundContext = scope;
+    const contextWrapper = function(this: {}) {
+      const oldContext = manager.current;
+      manager.current = boundContext;
+      const res = target.apply(this, arguments);
+      manager.current = oldContext;
+      return res;
+    };
+    Object.defineProperty(contextWrapper, 'length', {
+      enumerable: false,
+      configurable: true,
+      writable: false,
+      value: target.length,
+    });
+    /**
+     * It isnt possible to tell Typescript that contextWrapper is the same as T
+     * so we forced to cast as any here.
+     */
+    // tslint:disable-next-line:no-any
+    return contextWrapper as any;
+  }
+
+  /**
+   * By default, EventEmitter call their callback with their scope, which we do
+   *  not want, instead we will bind a specific scope to all callbacks that
+   *  go through it.
+   * @param target EventEmitter a instance of EventEmitter to patch
+   * @param scope the scope we want to bind
+   */
+  private bindEventEmitter<T extends EventEmitter>(
```

Comment:
The private bind methods should go in the base since they will be the same for any scope manager. If something is specific to Node, maybe there should be a specific base for Node, but I think functions, promises and event emitters apply to JavaScript in general.

Reply:
I don't agree with you on this one, i would be in favor of letting all implementation handle like they want each type. Maybe it would means have twice the same code but we are talking about <50 lines of code, which isn't enough to justify sharing it (at least for me).

Reply:
IMHO every single line of code should not be duplicated if it can be avoided. This is especially true for this case where every single implementation will have to repeat the exact same code which we know in advance.